### DISCUSSION TITLE
Accounting for "optimistic" provers who have more than n_f, but less than n_p elements

### DISCUSSION THREAD

**Discussion by amesgen** [cardano-scaling/alba#17] [General]
Citing from the introductory text of the [ALBA paper](https://eprint.iacr.org/2023/1655.pdf):

> More formally, let $R$ be a predicate. We explore succinct arguments of knowledge for a prover who knows a set $S_p$ of values that satisfy $R$ such that $|Sp| \ge n_p$ and wants to convince a verifier that $|S_p| > n_f$, where $n_f$ is somewhat smaller than $n_p$. Because $n_f < n_p$, the verifier obtains a lower bound approximation to the actual cardinality of $S_p$; hence we call this primitive an *Approximate Lower Bound Argument* or ALBA.
>
> [...]
>
> Let $\lambda$ be the parameter that controls soundness and completeness: the honest prover (who possesses a set of weight $n_p$) will fail with probability $2^{−\lambda}$ and the dishonest prover (who possesses a set of weight at most $n_f$) will succeed with, say, also probability $2^{-\lambda}$.

My understanding of this is that provers who have more than $n_f$, but less than $n_p$ elements satisfying $R$ (so $n_f < |S_p| < n_p$) are out-of-scope for the soundness and completeness properties. In particlar, under the confidence guaranteed by the security parameter, upon successfully verifying a proof, the verifier only learns that the prover has more than $n_f$ elements; but not something stronger, for example whether the prover actually had (close to) $n_p$ elements.

Therefore, it is natural to ask how easy it is for an "optimistic" prover, who only has $|S_p|$ elements with $n_f < |S_p| < n_p$, to still create a valid proof. The goal of this thread is to understand how such (adversarial) optimistic provers impact the ALBA parameterization choices for potential applications.

---

The most general strategy for such an optimistic prover is to simply try all possible proofs in order to find one that is valid. Lemma 3 (slightly modified) can be used to analyze this: In the scheme from Section 3.2.2 ("Generalization to small $n_p$"), there are $r \cdot d \cdot {|S_p|}^u$ possible proofs, and each one is valid with probability ${(1/n_p)}^u \cdot q$, so by the union bound, we get

$$ r \cdot {\left( \frac{|S_p|}{n_p} \right)}^u\cdot qd $$

as an upper bound on the probability that such a prover can create a valid proof.

---

As a concrete example, consider the parameters $n_p = 0.8 \cdot n = 720$ and $n_f = 0.2 \cdot n = 180$ where $n = 900$ as well as $\lambda = 128$. Based on Corollary 3, this yields $u = 70$, $d = 5567$, $q \approx 0.000893$ and $r = 128$.

![optimistic-alba-proving](https://github.com/user-attachments/assets/069cc98e-4d01-4f13-b10d-ec910c97c4c5)

Generally, the success probability increases exponentially the larger $S_p$ gets. We can see that the upper bounds on the success probability of the prover are non-negligible for values noticeably smaller than $n_p$, for example 1.8% for $|S_p| = 620$ or 0.005% for $|S_p| = 570$. For $|S_p| = 400$, we get a bound of 8.6e-16, which is quite low, but still orders of magnitude larger than the configured security parameter of $2^{-128}$ (rougly 3e-39).

If we imagine the context to be a voting scheme with $n = 900$ votes and a share of adversarial voters, where a valid ALBA proof signals that some quorum has been met, this has the following implications for example:

 - In addition to the possibility of the adversary getting $n_p$ votes, there also is the possibility of getting less than $n_p$ *and then* getting lucky so to produce a valid ALBA proof.
    - In particular, this becomes relevant when there is a possibility that the honest votes are split across e.g. two options, and the adversary can then vote for one or even both of them, in order to get closer to $n_p$ votes than they could with their votes alone. 
 - Suppose that honest participants only attempt to create an ALBA proof when they have at least some minimum number of votes. Suppose that an adversary is eligible for sufficiently many votes such that the honest participants do not reach this threshold. Then the adversary might still be able to create a valid ALBA proof, causing confusion at the very least.

The "easy" option to address problems like this (avoiding any complicated analysis) would be to make $n_f$ so large that just by the soundness of ALBA, one does not have to worry about optimistic provers. However, this might result in the ratio $n_p / n_f$ to be so small that ALBA becomes less or even unappealing for the particular application.

---

<details>
<summary>matplotlib script for the graph above</summary>

```python
#!/usr/bin/env nix-shell
#! nix-shell -i python3 --pure
#! nix-shell -p "python3.withPackages (ps: [ ps.matplotlib ps.scipy ])"
#! nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/5e0ca22929f3342b19569b21b2f3462f053e497b.tar.gz

from math import *
import matplotlib.pyplot as plt
import numpy as np

lam = 128  # security parameter
n = 900
n_p = int(0.8 * n)
n_f = int(0.2 * n)

print(f"n_p = {n_p}, n_f = {n_f}, λ = {lam}")

assert n_f < n_p

# From Corollary 3:
u = ceil((lam + log2(lam) + 5 - log2(log2(e))) / log2(n_p / n_f))

# Make sure we are in Case 1 in the proof of Corollary 3.
S_1_bot = n_p / (17**2 / (9 * log2(e)) * u**2) - 7 < 1
S_2_bot = n_p / (17**2 / (9 * log2(e)) * u**2) - 2 < 1
assert S_1_bot or S_2_bot

# With the parameters above, we are in Case 1 in the proof of Corollary 3 (or
# the "≤ λ²" column in Figure 1).
d = ceil(32 * log(12) * u)
q = 2 * log(12) / d
r = ceil(lam)

print(f"u = {u}, d = {d}, q = {q}, r = {r}")

optimistic_prover_success_bound = lambda v: r * (v / n_p) ** u * d * q


fig, ax = plt.subplots(figsize=(9, 6))
X = np.arange(n_f - 10, n_p + 1)
ax.set_title(f"Optimistic ALBA proving for n_p = {n_p}, n_f = {n_f}, λ = {lam}")
ax.set_xlabel("number of elements |S_p| the prover has access to")
ax.grid(True)

ax.set_yscale("log", base=10)
ax.plot(X, np.vectorize(optimistic_prover_success_bound)(X))
ax.set_ylabel("probability (upper bound) to create a valid proof")
ax.tick_params(axis="y")

secax = ax.secondary_yaxis(location="right")
secax.set_yscale("log", base=2)

ax.axvline(x=n_p, linestyle="--", color="red", label="n_p")
ax.axvline(x=n_f, linestyle="--", color="green", label="n_f")
ax.legend()
ax.set_ylim(top=1)

fig.savefig("optimistic-proving.png", dpi=300)
```

<details>

**Comment by ghost** [comment]
@amesgen I think your analysis is correct, and I wrongly conflated 3 things when I analysed the use of ALBA in the context of Peras: 1/ the expected quorum of 75% of committee size to issue a certificate, 2/ the meaning of $n_p$ and $n_f$ parameters, and 3/ the relationship they have to each other. 

There is no reason to set $n_f = S_p - n_p$, where $S_p$ is the expected total set of items, as I initially thought, e.g these are not the same as the expected adversarial stake assumption we usually take into account when analysing the safety of our stuff, and therefore there's no reason to use `n_p = 0.8 * stake` and `n_f = 0.2 * stake`. 

However, there's also no reason to assume that an adversary can have more items than their expected stake, e.g if we assume the adversary cannot control more than 1/3rd of the stake, then we by setting `n_f = 0.33 * stake` we should be "safe" because then ALBA guarantees that an adversary "cannot" make a certificate. If we set `n_p = 0.75 * stake`, our proof will be larger (eg. 60kB assuming each vote is roughly 500 bytes) but still manageable.

Should we want to increase adversarial stake then of course proof size would increase significantly, but then we could also increase $n_p$: with $n_p = 90$ and  $n_f = 40$, proof size is still at about 60kB.  Obviously, the close we get to 50% the more cumbersome it becomes.

**Reply by ghost** [reply]
Perhaps @tolikzinovyev or @rrtoledo could chime in and correct me if I am wrong

**Reply by amesgen** [reply]
Just a question about notation: When you write sth like `n_p = 0.8 * stake`, I am not sure what `stake` is referring to here. Maybe you instead mean `n_p = 0.8 * n` where `n` is the mean committee size of Peras?

---

> However, there's also no reason to assume that an adversary can have more items than their expected stake, e.g if we assume the adversary cannot control more than 1/3rd of the stake, then we by setting `n_f = 0.33 * stake` we should be "safe" because then ALBA guarantees that an adversary "cannot" make a certificate. If we set `n_p = 0.75 * stake`, our proof will be larger (eg. 60kB assuming each vote is roughly 500 bytes) but still manageable.

What about the possibilities mentioned above where the honest votes are split across two options, and the adversary votes for one or even both of them, or the option where the honest provers narrowly missed the quorum? AFAIU the quorum $\tau = 3/4 \cdot n$ was chosen specifically such that no two blocks can reach quorum at the same time for a 50% adversary due to a quorum intersection argument.

As a concrete example, consider your suggested parameterization (under the assumption that `n_f = 0.33 * stake` is supposed to mean `n_f = 0.33 * n` where `n = 900` is the mean committee size as in the remark above), so `n_p = 0.75 * n = 675` and `n_f = 1/3 * n = 300` (using `1/3` instead of `0.33` for nicer numbers) . Suppose that the honest votes (`2/3 * n = 600`) are split across two different blocks, so both have `1/3` of the votes. Now the adversary can vote for both of them with their `1/3 * n = 300` votes, such that each block has `1/3 * n + 1/3 * n = 2/3 * n = 600` votes. Because `τ = 3/4 * n = 675`, both blocks have not reached quorum, so everything fine until here.

However, the adversary can now act as an optimistic ALBA prover and use these 600 votes to still create valid ALBA proofs for both blocks with non-negligible probability (`0.046%` for each block individually). This seems problematic.

For reference, the same plot as above adapted for the parameters here:

![optimistic-proving](https://github.com/user-attachments/assets/d25de99b-15e2-454c-9c70-649cf78de20a)

> Should we want to increase adversarial stake

IIRC Peras is supposed to stay secure with close to 50% adversarial stake, so shouldn't ALBAs be parameterized accordingly?

E.g. in the example above, the probability for an optimistic ALBA prover to create a valid ALBA proof for one of the blocks when the adversary has 40% of the stake is already `16%`.

**Reply by tolikzinovyev** [reply]
@amesgen great questions! I think your analysis is correct. In your example where the adversary has <1/3 power and the honest votes are split equally among two choices (1/3 each), you need to set n_f = 2/3 n because the adversary can get 2/3 of votes for both choices but not >2/3 votes for any two choices, but n_p also must be 2/3 n because there are only 2/3 n honest people. So, this is challenging for ALBA, that requires n_p > n_f. Two practical options that I see is either making stronger assumptions (e.g. assume the adversary has 1/4 power) or be optimistic and set n_p = 4/5 n and fall back to vote concatination if this many votes have not been received. In principal, we can also have ALBA output certificates of variable size depending of |S_p|.

**Reply by ghost** [reply]
FWIW, I checked what the certificate size would be with $n_p = 900$ and $n_f = 490$ for an average committee size of 1000, and a single vote of 500 bytes, and I got the following results:

```
% cabal run alba -- prove --n-p 900 --n-f 490 --size 900 --len 500 --security 128
Generating random items
Generating proof Options {size = 900, len = 500, params = Params {λ_sec = 128, λ_rel = 128, n_p = 900, n_f = 490}, input = Nothing, output = "proof.alba"}
Written proof to 'proof.alba' (81298 bytes, 18 retries)
```

Not great, but still in the range of a block size and there's work underway on ALBA certificates, possibly combined with ZK techniques, to reduce that size. And of course the votes themselves could potentially be improved. 

Moreover, in the case of Peras the scenario you outlined @amesgen is extremely unlikely: 

1. the cutoff parameter $L$ for selecting block to vote for is meant to ensure that honest votes will have a high probability of having converged on the same chain,
2. the final version of Peras which is still being worked on by the researchers will include a pre-agreement BFT functionality in order to guarantee exactly that.

It seems to me that not all hope is lost and we could adjust the $n_p$ and $n_f$ to be "nicer" and produce smaller proofs.

I will adjust Peras' report according to your comment, and I also created #18 to incorporate the information and model you provided @amesgen  into Peras' website.

**Reply by amesgen** [reply]
Thanks for the replies!

> It seems to me that not all hope is lost and we could adjust the $n_p$ and $n_f$ to be "nicer" and produce smaller proofs.

Yes, though it will definitely require careful analysis to be sure that situations that would enable optimistic provers to succeed can only arise with sufficiently low probability :+1:

> I will adjust Peras' report according to your comment, and I also created https://github.com/cardano-scaling/alba/issues/18 to incorporate the information and model you provided @amesgen into Peras' website.

Sounds great, also happy to help/review :+1:

**Reply by ghost** [reply]
> be sure that situations that would enable optimistic provers to succeed can only arise with sufficiently low probability

I would like also to emphasise that while the papers produce theoretical arguments based on <50% adversarial stake, the numbers used in practice seem to actually target a lower amount of stake. For example, following the computations from  [this paper](https://eprint.iacr.org/2022/1571) about settlement time, the 2160 blocks security parameter represents the number of blocks it takes to have less than $2^{-60}$ probability that an adversary controlling less than 35% stake with $10^{18}$ grinding power will be able to revert a transaction.

It seems we don't have a consistent _practical_ definition of what _Secure_ really means, and what adversarial stake we really take into account, in _practice_.

What could be valuable would be to frame analysis of the ALBA protocol not in terms of probabilities to find a proof, but in terms of computing power to find one with certainty. This would make it possible to recast the security argument of ALBA in the same terms as the one for underlying Praos chain for example, and unify the various probabilities and exponents under a shared terminology.

> happy to help/review

That would be great, will tag you once I have the PR ready.

**Reply by amesgen** [reply]
Yeah, at least as soon as significant nonce grinding comes into the picture, ones needs to consider less adversarial stake to still get a secure system :+1: I agree that it is unsatisfying that there don't seem to be concrete worst case values for adversarial stake/grinding power that one can use; the approach that we have taken so far is to calculate it for a range of adversaries, and if the resulting probabilities are not obviously way too high/low, ask other people (researchers/product people) to make the call on that.

I guess my point was that there is a tradeoff between picking "nice" $n_p$/$n_f$ (that e.g. result in small proofs/certificates) and having to do a more complicated analysis that this is actually secure when assuming certain adversarial resources (stake/grinding power/computational power).

**Reply by ghost** [reply]
Agreed, and again thanks for the great questions and pointing out that mistake of mine!

**Comment by ghost** [comment]
Using your formula for probability of finding a proof and some assumptions on hashing power, I found out that with $n_p = 900$, $n_f = 350$, a prover with 490 votes and $10^9$ CPUs would still need $10^7$ seconds to find a proof on average, or 115 days, which might probably be good enough for our use cases (but that would require more careful analysis of the opportunity window for an attacker to produce such an adversarial certificate).

 I have put up a  PR (#22), it would be great if you could check my computations are sound :)

**Reply by tolikzinovyev** [reply]
How is the proving time for the adversary so large? Wouldn't the adversary be able to run DFS just like the honest prover? It should be fast if not faster.

**Reply by ghost** [reply]
You tell me, you invented ALBA :) I just reused @amesgen's computations above, factoring in estimations for hashing power to translate probabilities to time.

**Comment by rrtoledo** [comment]
I had quite a few discussions on what nf and np with regards to our Peras protocol parameters, more specifically the quorum size which equals 75% of the comittee stake.

Alba is a proof of lower bound (I have seen at least nf elements satisfying a certain relation R). What matters "first" is the choice of nf, then we can consider np value which defines both when we can be sure to produce a proof and the proof size.

An opportunistic prover, who hasn't seen at least np elements, in my mind is not adversarial: he is still computing a proof that he has seen nf elements, but the probability of generating such proof is lower. The real adversary is someone trying to create a certificate with fewer than nf elements. I do not believe we have to worry about race conditions here either, the biggest worry of Peras is to always produce a certificate.

To come back to numbers, we will need to set nf to the quorum size, i.e. 75% of the commitee stake. This is to prevent an attack on Peras, the intersection attack, where the adversary can boost several blocks and in fine create forks. In practice, we may be able to lower this number to 67% if we consider a 35% adversary which as @abailly-iohk mentioned is considered elsewhere. 
As for np, we would like to have a np high enough so that the proof size is as low as possible but lowering np should decrease the proving time. What we could for np then is choose the smallest value such that the certificate fits in a block, of course given we can use a whole block. 

I have so far dismissed the third main Alba parameter, the security parameter lambda. It is usually set to have 128 bit of security. There might be opportunities to lower it depending on its value in the Peras protocol.

**Reply by ghost** [reply]
This seems to mean that $n_p/n_f$ ratio will be quite small, lower than any of the values that are considered in the original paper (eg. in the comparison table on p.37). With $n_p = 100$ and $n_f = 66$ that's 1.5 which equates to 241 votes according to the table. If a single vote has the size given in the Peras tech report (eg. 710 bytes) that's about 170kB. I understood you had ideas on how to "compress" that, is this explained somewhere?

**Reply by ghost** [reply]
Of course, there might a voting scheme were individual votes would be smaller than 710 bytes

**Reply by rrtoledo** [reply]
We are looking at different solutions:
- Lowering the parameter lambda
- Doing 2 rounds of Alba: one round where nf is set at 75% of the stake to get the desired property, and another round on these certificates with a low nf parameter to compress it further
- Doing 2 rounds of Alba: We create a certificate out of the Peras votes and broadcast them, then uses these as "votes" for a second round of Alba. There is a variant from Nicolas where we set nf=1 at the second round 

Out of scope at the momet:
- Alba + SNARK: It is out of scope right now as it would need more time for implementation and would be application dependent.

**Reply by amesgen** [reply]
The small $n_p/n_f$ ratio with $n_f = \tau n$ (the quorum size, where $n$ is the mean committee size) resulting in large certificates is indeed what prompted me to [raise this on Discord](https://discord.com/channels/826816523368005654/1247688695113449523/1273702948844601395) back then in the first place :+1:

> As for np, we would like to have a np high enough so that the proof size is as low as possible but lowering np should decrease the proving time. What we could for np then is choose the smallest value such that the certificate fits in a block, of course given we can use a whole block.

There is another factor here, favoring smaller values for $n_p$: Namely, $n_p$ needs to be small enough that a narrow quorum can actually be certified.[^1] Concretely, if one sets $n_f = \tau n$, we can have a quorum by virtue of observing a bit more than $\tau n = n_f$ many votes, e.g. $n_f + 1$ or $n_f + 10$ many votes. The larger $n_p$ is chosen, the less likely it gets that one will actually be able to create a corresponding valid ALBA proof.

I plotted this for the proposed Peras mean committee size of $n=900$ (so $\tau n = 3/4 \cdot n = 675$):

![alba](https://github.com/user-attachments/assets/9b5390f0-9929-4b84-b2df-6c9be6791573)

This is looking rather grim, the certificate sizes are much larger than the current Cardano mainnet block limit in all cases (even when choosing $n_p$ to be significantly larger than the mean committee size, which has its own problems). In addition, the probability (note that these are conservative upper bounds; the actual probability might be even lower) to certify narrow quorums drops rapidly when choosing larger $n_p$.

Also, as expected, ALBA is actually less efficient than simply concatenating $n=900$ votes when $n_p$ small (i.e. when the horizontal red line is below the blue line in the top diagram).

Repeating the same and decreasing $\lambda$ from $128$ to $64$ doesn't really change the picture:

<details>
<summary>Click to see plot for $\lambda=64$</summary>

![alba](https://github.com/user-attachments/assets/7a306ee6-a80f-4fd8-98cb-6c16269edd8b)

</details>

Here, the certificate size is smaller than the current Cardano mainnet block limit for $n_p \approx 1020$ (eyeballing), but there, the probability to create a valid certificate out of $n=900$ votes (the mean committee size) is already smaller than $10^{-4}$ (see the brown line), so this isn't really workable.

---

<details>
<summary>matplotlib script for the plots above</summary>

```python
#!/usr/bin/env nix-shell
#! nix-shell -i python3 --pure
#! nix-shell -p "python3.withPackages (ps: [ ps.matplotlib])"
#! nix-shell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/5e0ca22929f3342b19569b21b2f3462f053e497b.tar.gz

from math import *
import matplotlib.pyplot as plt
import numpy as np


def alba_params(n_p, n_f, lam):
    assert n_f < n_p

    # From Corollary 3:
    u = ceil((lam + log2(lam) + 5 - log2(log2(e))) / log2(n_p / n_f))

    # Make sure we are in Case 1 in the proof of Corollary 3.
    S_1_bot = n_p / (17**2 / (9 * log2(e)) * u**2) - 7 < 1
    S_2_bot = n_p / (17**2 / (9 * log2(e)) * u**2) - 2 < 1
    assert S_1_bot or S_2_bot

    # With the parameters above, we are in Case 1 in the proof of Corollary 3 (or
    # the "≤ λ²" column in Figure 1).
    d = ceil(32 * log(12) * u)
    q = 2 * log(12) / d
    r = ceil(lam)

    return u, d, q, r


def optimistic_prover_success_bound(n_p, n_f, lam):
    u, d, q, r = alba_params(n_p, n_f, lam)
    return lambda v: r * (v / n_p) ** u * d * q


lam = 128  # security parameter
n = 900
n_f = int(0.75 * n)

bytes_per_vote = 710

max_block_size = 88 * 1024

fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(9, 12))
X = np.arange(n_f + 1, n + 300)
ax1.set_title(f"ALBA proof size for n_f = {n_f} = 3/4 * {n}, λ = {lam}\n")
ax1.set_xlabel("n_p")
ax1.grid(True)

ax1.plot(X, np.vectorize(lambda n_p: alba_params(n_p, n_f, lam)[0])(X))
ax1.set_ylabel("number of elements in ALBA proof")
ax1.set_yscale("log")

axbytes = ax1.secondary_yaxis(location="right", functions=(lambda u: bytes_per_vote * u, lambda b: b / bytes_per_vote))
axbytes.set_ylabel("ALBA proof size [B]\nassuming vote size of 710B")

axratio = ax1.secondary_xaxis(location="top", functions=(lambda n_p: n_p / n_f, lambda r: r * n_f))
axratio.set_xlabel("n_p / n_f")

ax1.axvline(
    x=n,
    color="red",
    label=f"mean committee size n = {n}",
)
ax1.axhline(
    y=n,
    color="red",
)
ax1.axhline(
    y=max_block_size / bytes_per_vote,
    color="green",
    label=f"current Cardano mainnet max block body size ({max_block_size}) [B]",
)

ax1.legend()

ax2.set_title(f"Upper bounds on probability to certify a narrow quorum")
ax2.set_xlabel("n_p")
ax2.grid(True)

for off in [1, 10, 25, 50, 100, n - n_f]:
    v = n_f + off
    X = np.arange(v, n + 300)
    ax2.plot(
        X,
        np.vectorize(lambda n_p: optimistic_prover_success_bound(n_p, n_f, lam)(v))(X),
        label=f"|S_p| = n_f + {off}" if v != n else f"|S_p| = n = {n}\n(for reference)",
    )
ax2.set_ylabel("upper bound on the probability to create an ALBA proof\nwith the given number |S_p| of elements")
ax2.set_yscale("log")
ax2.set_ylim(top=1)
ax2.legend()

fig.savefig("alba.png", dpi=300, bbox_inches="tight")
```

<details>

[^1]: @tolikzinovyev mentions in https://github.com/cardano-scaling/alba/discussions/17#discussioncomment-10627410 that one could fall back to vote concatenation in this case, but this would then be yet another risk factor that has to be analyzed.

**Reply by ghost** [reply]
As already discussed, I will update the Peras report to suggest the initial ALBA picture was way too optimistic and we need more work to be able to build workable certificates. Note that we could resort to use Mithril and BLS multisignatures, for which we have some experience but this entails its own set of problems: how to share the keys, and computation time which is heavier.

**Reply by tolikzinovyev** [reply]
It seems that what Peras needs is exactly what the decentralized ALBA provides: proving that a set of parties with sufficient total weight (3/4 stake) vote for the same message (block). Also, I might be wrong, but @amesgen's calculations look somewhat inaccurate. In reality, the actual committee size in a simple lottery will be different from its expected size 900, 3/4 of the stake is likely to have significantly more than 675 seats in the committee, and 100% of the stake is likely to have much fewer than 900 seats in the committee. For this reason, setting n_f = 675, n_p = 900 for ALBA isn't valid and the ratio n_p / n_f needs to become even smaller. But same applies to the simple lottery (with concatenation): the threshold (and so does the proof size) will need to be significantly larger than 675.

I would like to shamelessly advertise my recent report I shared on slack about decentralized ALBA. :)  It has graphs for the tradeoff between expected committee size (communication complexity mu) and the proof size u. In particular, they show that Telescope ALBA is always superior to the simple lottery in terms of the proof size. I can also share the code that generates those graphs if you want to try custom parameters.

Relevant to this, I think the two round voting solution proposed by Peter Gazi seems to be the most elegant solution. The first round can use the simple lottery to establish a _unique_ block without regard for certificate size, while the second round would use ALBA with a much smaller n_f equal to the adversarial stake to generate a small certificate.

Let me know your thoughts. I'm happy to discuss this further.

**Reply by ghost** [reply]
I had a vague intuition a couple of months ago the decentralised ALBA could be a good fit for Peras because it includes the weight and therefore the voting into the ALBA scheme, while the centralised construction with single votes require weight to _also_ be in the votes. It would be great to investigate this possibility further.

**Reply by amesgen** [reply]
> Also, I might be wrong, but @amesgen's calculations look somewhat inaccurate. In reality, the actual committee size in a simple lottery will be different from its expected size 900, 3/4 of the stake is likely to have significantly more than 675 seats in the committee, and 100% of the stake is likely to have much fewer than 900 seats in the committee. For this reason, setting n_f = 675, n_p = 900 for ALBA isn't valid and the ratio n_p / n_f needs to become even smaller. But same applies to the simple lottery (with concatenation): the threshold (and so does the proof size) will need to be significantly larger than 675.

Oh yeah, that is definitely simplified in the discussion above (using unweighted ALBA) as in reality, different participants in a Peras committee might have different (not equally distributed) stake :+1: An easy workaround, assuming that the Peras committee is chosen via an Algorand-like sortition, is to let individual participants potentially vote several times (according to their sample from a binomial); but that is basically the same as (or worse than) using a weighted ALBA variant directly AFAIU.

> Relevant to this, I think the two round voting solution proposed by Peter Gazi seems to be the most elegant solution. The first round can use the simple lottery to establish a unique block without regard for certificate size, while the second round would use ALBA with a much smaller n_f equal to the adversarial stake to generate a small certificate.

FTR I don't know the details as I am probably not in the relevant channels on Slack (which is fine, I only started looking into ALBA on the side while reading the Peras reports/CIP and getting stuck on the parameterization), but that sounds cool, looking forward to it :+1:

**Reply by tolikzinovyev** [reply]
@abailly-iohk I'm not sure what you mean by "including the weight". Both centralized and decentralized alba can be weighted or unweighted. Weighted means that data elements can have integer weights and you are proving knowledge of element of sufficient total weight.

@amesgen You are right that decentralized alba is just simple lottery (algorand-like sortition) plus centralized unweighted alba. But what I was trying to say is that the analysis of security and reliability, and therefore parameter derivation, is more complicated. Good news is that it has already been done in the alba paper and my report on slack has nice graphs.

By the way, I'm also confused why in https://peras.cardano-scaling.org/docs/reports/tech-report-2/#no-honest-quorum-in-round the 3/4 quorum is fixed while the fraction of adversarial stake f is varied. In reality, if f is small, we also need a smaller quorum. We basically need to avoid the situation where honest votes are split equally among two different blocks and the adversary adds his votes to both sets to certify both blocks. So we need the quorum size tau > (1 - f) / 2 + f = (1 + f) / 2. If f is 50% this gives quorum size 75%, but if f is smaller, the quorum can be smaller. When using decentralized alba, we therefore need n_f = (1 + f) / 2.

**Reply by amesgen** [reply]
> You are right that decentralized alba is just simple lottery (algorand-like sortition) plus centralized unweighted alba. But what I was trying to say is that the analysis of security and reliability, and therefore parameter derivation, is more complicated. Good news is that it has already been done in the alba paper and my report on slack has nice graphs.

That makes sense, I just wanted to get a feeling for how the parameters turn out in the centralized unweighted setting when setting $n_f = \tau n$ to confirm that the results aren't appealing, and my intuition is that using a variant with weights/decentralization won't make the numbers *better*. However, when actually constructively choosing ALBA parameters in a setting where $n_f$ can be smaller (e.g. the two-round scheme sketched above), e.g. your report sounds like exactly what should be used.

> By the way, I'm also confused why in https://peras.cardano-scaling.org/docs/reports/tech-report-2/#no-honest-quorum-in-round the 3/4 quorum is fixed while the fraction of adversarial stake f is varied. In reality, if f is small, we also need a smaller quorum. We basically need to avoid the situation where honest votes are split equally among two different blocks and the adversary adds his votes to both sets to certify both blocks. So we need the quorum size tau > (1 - f) / 2 + f = (1 + f) / 2. If f is 50% this gives quorum size 75%, but if f is smaller, the quorum can be smaller.

I am not the author, but that's how I understood it: Peras (at least in the "pre-alpha version" of Peras, ie without any pre-agreement) only provides *optimistic* fast settlement. We assume a worst case adversarial stake ratio of $f=1/2$, so we need to set the quorum $\tau$ to at least $3/4$. Now:

 - An adversary who has more than $1/4=1-\tau$ of the stake but less than $f=1/2$ can effectively disable Peras all the time by preventing quorum formation (by simply abstaining), but they can't compromise the security of the underlying protocol as they still have less than $f=1/2$ of the stake.
 - An adversary who has less than $1/4=1-\tau$ of the stake can only prevent honest quorum formation with a small probability. The purpose of this section in the Peras report is to quantify how likely this is going to be the case, in order to choose a sufficiently large committee size to make sure that "weak" adversaries (with e.g. 10% stake) can't disable Peras.

**Reply by tolikzinovyev** [reply]
Thanks for explanation, it makes sense about optimistic fast settlement. So the problem statement is that we want the adversary to successfully certify two different blocks only with tiny probability even if adversarial stake is 50% (or close to 50), but the protocol should provide fast settlement when adversarial stake is actually much lower? But then I'm confused why https://peras.cardano-scaling.org/docs/reports/tech-report-2/#adversarial-quorum assumes f to be smaller than 25%. It doesn't solve the problem of safety against 50% adversarial stake. Also the threshold tau would need to be higher than 3/4 * expected committee size as there is deviation due to lottery randomness.

The decentralized alba can give proof size as small as ≈ (lambda + log(lambda) + 5) / log(n_p / n_f). For lambda = 128, n_p = 100% stake, n_f = 75% stake, this is 337. Of course, the communication complexity would be large in this case. There is a tradeoff between the two.

**Reply by amesgen** [reply]
> So the problem statement is that we want the adversary to successfully certify two different blocks only with tiny probability even if adversarial stake is 50% (or close to 50), but the protocol should provide fast settlement when adversarial stake is actually much lower?

That's my understanding (where "close to 50%" is a bit fuzzy), maybe someone from the Peras team can confirm this.

> But then I'm confused why https://peras.cardano-scaling.org/docs/reports/tech-report-2/#adversarial-quorum assumes f to be smaller than 25%. It doesn't solve the problem of safety against 50% adversarial stake.

Yeah, I agree that it would be useful for this plot to include higher adversarial stake ratios.

> The decentralized alba can give proof size as small as ≈ (lambda + log(lambda) + 5) / log(n_p / n_f). For lambda = 128, n_p = 100% stake, n_f = 75% stake, this is 337. Of course, the communication complexity would be large in this case. There is a tradeoff between the two.

Hmm, but setting n_p = 100% stake would then probably make "narrow quorums" unprovable/uncertifiable as mentioned above, right? I.e. suppose that there is a quorum for a block by 80% of the stake. What would be the completeness error to create a corresponding ALBA proof?

**Reply by tolikzinovyev** [reply]
> What would be the completeness error to create a corresponding ALBA proof?

It would be pretty close to 0. To get completeness error 1/2 with n_p / n_f = 80/75, you would need to make proof size 1483.
```
alba-tradeoff/code$ cargo run --release -- --lamsec=128 --lamrel=1 --ratio=1.066667 --scheme=telescope-bounded calc --target-mu=1000000
    Finished `release` profile [optimized] target(s) in 0.03s
     Running `target/release/decentralized-alba-params --lamsec=128 --lamrel=1 --ratio=1.066667 --scheme=telescope-bounded calc --target-mu=1000000`
u: 1483 mu: 980301
```

I will share my calculation code in https://github.com/tolikzinovyev/alba-tradeoff with you and @abailly-iohk if you want to try different parameters.

**Reply by tolikzinovyev** [reply]
For simple lottery it's much worse...
```
alba-tradeoff/code$ cargo run --release -- --lamsec=128 --lamrel=1 --ratio=1.066667 --scheme=simple-lottery calc 
    Finished `release` profile [optimized] target(s) in 0.10s
     Running `target/release/decentralized-alba-params --lamsec=128 --lamrel=1 --ratio=1.066667 --scheme=simple-lottery calc`
u: 51380 mu: 51647.34936883665
```
